### DISCUSSION TITLE
Support repeating multiple times per day

### DISCUSSION THREAD

**Discussion by jonblack** [iSoron/uhabits#121] [Feature Requests]
A really good use case for this is taking medication, which sometimes needs to happen multiple times a day.

**Comment by iSoron** [comment]
Thank you for the suggestion. This is an interesting feature, but I am not sure how to implement it and keep the interface simple. I though about two alternative solutions. Neither is perfect, but they might work well enough.

The first is to create multiple habits (take medicine at breakfast, take medicine at night). This can be done already, with the drawback that you cannot see unified statistics / widgets for both. Features #51 and #45 could help.

The other solution would be using numerical habits #42 (how many times have you taken your medicine today?). We could allow numerical habits to have multiple reminders throughout the day.

If anyone has better ideas, feel free to share.

**Reply by herrbrixter** [reply]
I do this. I just have 2 habits: Take mornings meds, Take evening meds.

**Comment by jonblack** [comment]
> This is an interesting feature, but I am not sure how to implement it and keep the interface simple.

I'm assuming you mean the "cross" and "tick" part on the main screen. For a "multiple day" habit you could require a long press which shows a popup of some sort. I agree it's not trivial.

Without the popup you could use another symbol (e.g. ~) or an perhaps an orange tick to show that the habit has been partially met.

> The first is to create multiple habits.

This is a nice workaround for now but it's indeed a shame about the unified statistics.

> The other solution would be using numerical habits.

This could work. Multiple reminders would be (at least for me) required because timing is very important :smile:

**Comment by LuccoJ** [comment]
"Numerical habits" in the sense of having a single question being asked "How many times have you ... today?" might not be ideal, because you'd probably rather tell the app directly each time you (for instance) take your medication, right when you do.

Perhaps you could have "numerical habits" in the sense that, by default, habits would assume you want to perform them once a day, but you could, instead, specify that a given habit has a goal of 3 times a day, and then, when you "tick" that habit, instead of a tickmark showing up, a progress bar/circle/number/something would show up.

**Comment by SherylHohman** [comment]
I also would like to see this incorporated. It is something I think about quite a bit, actually.

I would use numerical counts alongside "mini habits".  
Perhaps I set a mini habit to do 10 minutes of something. 
As long as I do 10 min, my mini-habit goal is met. 
On some days I go above and beyond - say 30 or 60 minutes. 
Then while my minimum is 10, anytime I do at least 10, I've met my goal and am happy. 
However, I'd also like to see what pattern I follow when I do do more. 
If fact, when I see that I've consistently performed 20-60 min most days for a few months, then I might take that as indication to increase my minimum.. Or, as a tracker, I might just like to "get credit" for what I actually do. Mini-habit completion is used for constructing a habit effortlessly, which should be ridiculously easy to accomplish, and consist of a simple true/false statement.
The additional metric is actually a track of measuring work performed, which is quite different from habit tracking. Indeed, I am more likely to work in spurts: nothing, nothing, nothing, 12hr, 10hr, nothing... Which is exactly why I need an app to motivate me to create mini-habits.
But the measuring side of me really wants credit when I do more than the minimum. 
Spurts on top of mini-habits is a wonderful middle ground that I'm after.

To this end, I've used a couple different methods.
I currently have multiple entries for certain items 
- medication/vitamins: 
  - am (sometimes happens in am, sometimes late afternoon) 
  - pm (often gets skipped)
- workouts:
  - weights-any, 
  - weights 30min+, 
  - weights 60min+
  - weights 2hr+ , 
  - [the same set repeated for cardio], and 
  - "any calisthenic (ie 15 lounges, 10 push-ups), hourly",
    _for at least hours (4x) in the day_  
    (kind of like a pomodoro thing, and prompts me "move", even when I don't otherwise exercise that day)

You can see I currently employ both philosophies mentioned above:
I see the value of setting a "minimum of 1x", but tracking the number of times I actually performed the activity. (minimum cardio and weights is Any attempt (mini habit), but count in 30min blocks for tracking and stats) (I'm likely to exercise at least 2hrs when I _do_ exercise, but I need consistency.  I've long tracked total exercise performed, and I want to continue doing so. I also want to track habitualness now)
I see the value in having a "minimum of 4x", so each time I do the activity, the count is incremented, but it only gets marked off if I do 4 or more times.

I can see the value of having stats combining am and pm medication on a single chart. Although, I might prefer it showing  line graphs: 1 for am and another for pm. or a plot showing deviation from expected times. I'm not really sure how I'd want that data represented, actually.. 
I find that I'm more likely to take am vitamins/medication, than pm, even when both are required.Then again, taking "am" in the morning doesn't actually occur until late afternoon at times! So perhaps a plot of the actual times I take it, would be useful.
I have ideas on how I'd like workout data, or other kinds of output-related habits (writing, coding, etc). Not so much on medication/vitamins. Curious what you'd like to see.

---

I like the idea of using a number, rather than a checkmark, if it is performed more than once in a day, and turning from the light color to the darker (minimum completed color) once the "minimum" requirement has been met.

For example: if the  minimum to fulfill the habit is 4, then each time the habit is performed, count 1, 2, 3, 4, 5, .. At 0-3 the habit would remain grey. At 4 it could turn darker. In this case, it may also turn into a checkmark.
On the other hand, if the minimum is 1: Once the habit is performed, a darkened checkmark is placed. For each additional repeat, it counts up 2, 3, 4.. , in the darkened color.

The only drawback with this is that a completed habits will appear as a combination of numbers and checkmarks, which is visually less uniform.  
Likewise, incomplete habits could appear as a combo of x's and numbers, which might kinda blend together  
The visual identifier will be light vs dark for complete vs incomplete - As such, it might make sense to increase the contrast between light and dark. The difference in contrast should become the predominant visual indicator (overshadowing the x's, numbers, and checkmarks jumble).

Another way to make a distinction when incorporating numbers is to change the x's to 0's or O's  
O's will stand out more than x's, which would blend in with numbers.  
This would make habits that were not touched that day pop out easier during a quick visual scan.

I'd be interested to see how this additional data would be plotted !

**Comment by askmydog** [comment]
I want to voice my support of this issue. The best part of the app is being able to press that little widget button and see my habit strength go up. Your habit strength algorithm is awesome, BTW.

The "problem" is that I don't get credit if I go past the numerical goal that I set. So if I set a goal to read 2 articles per day, I get to tick off my goal when I read two articles, but if I read more than two, the only reward I get is just knowing I'm doing what I'm supposed to do (lame). 

Besides this, for some goals, I'm more concerned about how many individual times per week than how many days per week I performed the task. For instance, I might want to read 10 articles per week, but I might read 4 articles on one day and only 1 day the rest of the days. 

Setting a goal for 2 articles per day doesn't account for that variation. Setting a goal of 10 articles per week makes it more difficult to keep track of how many articles I've read.  Plus, if I have that many individual tasks over that long of a period risks, I invariably wait until the last minute to read all 10 articles, even if I set reminders. 

Like SherylHohman suggested, I wouldn't have any issues with replacing the x and check mark in the widget with a counter, and have it turn colored when the goal is reached. If the goal is a simple yes/no (ie, run 3 days per week), the user could just keep the check mark. 

Likewise, on the main screen of the app, it could display the number of times the habit was checked for each day rather than a check mark, and only turn colored when the goal was reached. The number could stay colored if the user maintains a steak, and have an extra circle around the number if the users goes above the goal. 

The habit strength algorithm shouldn't change, but maybe there should be a bar on the graph indicating the number of times the task was competed that week.  

BTW, the habit algorithm is what sets your app apart from the competition, because the fact that the habit can degrade over time is very motivating, way more than just trying to achieve a certain number times by a certain date like the other apps. 

I hope this helps, and keep up the good work!

**Comment by packetmagic** [comment]
I would love to have this feature. There are tasks that repeat a different number of times everyday. Being able to easily count the number of times they are completed would be wonderful. Perhaps as suggested before you could use a number instead of a check mark. You could use a double tap to increase the number and a tap and hold to decrease it.

**Comment by imme-emosol** [comment]
edit: made a comment to quickly. Just read the comments already made :/ sorry, my bad.
Could a distinction between double & single taps be made .. ?
single tap > log one entry for habit (with current timestamp)
double tap (or perhaps 'long press') > remove previous entry for habit

**Comment by imme-emosol** [comment]
To add even more body to this; I would like to be able to keep track of a certain habit, not to reach a goal but just to get a notion of my current repetition of it.

In my specific case, I was hoping to be able to count the amount of times I drink coffee over time.

**Comment by RichyHBM** [comment]
I'm in the scenario where I would like to have X tasks per day with random reminders, potentially being able to specify the window in which they could happen.

For example, between 9am and 8pm, have 5 reminders to check-in that I haven't had a snack

**Comment by petkahl** [comment]
I too would really like this feature. There are two versions that would be useful to me:

- Do habit X times per day
- Track how many times I do a habit per day. 

The second one isn't necessarily on/off. If I do the habit in a day, it's "done". But I'd like to know that, for example, I noticed myself procrastinating and got myself back on track 6 times today, but only twice yesterday.

I'd go for a yellow or grey tick mark for "partial" habits, and in those cases, don't count them as complete (for the purpose of hiding completed habits) until they're done the requisite number of times, or ever in the case of open-ended.

**Comment by amaramrahul** [comment]
I like the option for multiple reminders in some other apps such as Habit Bull / TickTick, wherein the task and reminders are not related. For ex. I might want to drink 3 glasses of water in a day, but I would like to be reminded at 6 different times a day.

This really is the only thing that is avoiding me from using Loop Habit Tracker and having such a feature would make it so much more useful :).

**Comment by animaone** [comment]
Also there should be an option to ring an alarm until the phone owner do the task. I would call this 'torture mode', I think 'torture mode' is the perfect word for this. Enable torture mode in a per-task basis or enable it globally in the settings. :P, the phone will torture you with noise and shame until you press the done button.

**Reply by sh0nuff** [reply]
I'd recommend posting this as a separate feature, as it's not really the same as the original suggestion! =)

**Comment by sh0nuff** [comment]
> Also there should be an option to ring an alarm until the phone owner do the task. I would call this 'torture mode', I think 'torture mode' is the perfect word for this. Enable torture mode in a per-task basis or enable it globally in the settings. :P, the phone will torture you with noise and shame until you press the done button.
I'd love to see this feature as well. I've been thinking about how to implement it, and had a couple ideas. 

In order to make it a little easier to configure, it could be only able to be triggered through the widget. For example, drink water every day could have a target of 3x in 24h, and each time I tap the widget for this habit, it lights up the first time I tap, but each extra time it adds a number in the circle of how many times I tap it. 
Another way would be to also have this same behavior in the task / app list as well, and allow for multiple taps for that day w/ a number count, but require a long press to call up a dropdown where users could reset to 0 if they tapped in error, or slide up and down to move in increments (0,1,2,3 etc)
